djangosaml2idp/processors.py

```diff
@@ -47,14 +49,20 @@ def _get_nameid_opaque(cls, user_id: str, salt: bytes = b'', *args, **kwargs) ->
         return opaque.hexdigest()
 
     @classmethod
-    def get_nameid_persistent(cls, user_id: str, user: User, sp_entityid: str = '', idp_entityid: str = '') -> str:
+    def get_nameid_persistent(cls, user_id: str, sp: ServiceProvider, user: settings.AUTH_USER_MODEL) -> str:
         """ Get PersistentID in TransientID format
             see: http://software.internet2.edu/eduperson/internet2-mace-dir-eduperson-201602.html#eduPersonTargetedID
         """
-        return '!'.join([idp_entityid, sp_entityid, cls._get_nameid_opaque(user_id, salt=str(user.pk).encode())])
+        if PersistentId.objects.filter(sp=sp, user=user).exists():
```

Comment:
by using .filter().first() you only do one query which either yields None, or the actual item if it exists.

Reply:
I also think we could just use `.get_or_create` here and this be basically a one liner. https://docs.djangoproject.com/en/3.0/ref/models/querysets/#get-or-create

Reply:
Done, good idea!